File: csharp/Resources/ElevateResource.cs
using System.Net.Http;
using System.Threading.Tasks;
using Newtonsoft.Json.Linq;

namespace Authdog
{
    /// <summary>
    /// Privileged-access elevate Wave 3 operations
    /// </summary>
    public class ElevateResource
    {
        private readonly AuthdogClient _client;

        public ElevateResource(AuthdogClient client)
        {
            _client = client;
        }

        public Task<JObject> ActivateGrantAsync(string tenantId, string environmentId, string grantId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/elevate/access-grants/{grantId}/activate",
                body);

        public JObject ActivateGrant(string tenantId, string environmentId, string grantId, object body) =>
            ActivateGrantAsync(tenantId, environmentId, grantId, body).GetAwaiter().GetResult();

        public Task<JObject> RevokeGrantAsync(string tenantId, string environmentId, string grantId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/elevate/access-grants/{grantId}/revoke",
                body);

        public JObject RevokeGrant(string tenantId, string environmentId, string grantId, object body) =>
            RevokeGrantAsync(tenantId, environmentId, grantId, body).GetAwaiter().GetResult();

        public Task<JObject> ListRequestsAsync(string tenantId, string environmentId, string? status = null) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"{AuthdogClient.Env(tenantId, environmentId)}/elevate/access-requests",
                query: AuthdogClient.Params(("status", status)));

        public JObject ListRequests(string tenantId, string environmentId, string? status = null) =>
            ListRequestsAsync(tenantId, environmentId, status).GetAwaiter().GetResult();

        public Task<JObject> CreateRequestAsync(string tenantId, string environmentId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/elevate/access-requests",
                body);

        public JObject CreateRequest(string tenantId, string environmentId, object body) =>
            CreateRequestAsync(tenantId, environmentId, body).GetAwaiter().GetResult();

        public Task<JObject> GetRequestAsync(string tenantId, string environmentId, string requestId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"{AuthdogClient.Env(tenantId, environmentId)}/elevate/access-requests/{requestId}");

        public JObject GetRequest(string tenantId, string environmentId, string requestId) =>
            GetRequestAsync(tenantId, environmentId, requestId).GetAwaiter().GetResult();

        public Task<JObject> ApproveRequestAsync(string tenantId, string environmentId, string requestId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/elevate/access-requests/{requestId}/approve",
                body);

        public JObject ApproveRequest(string tenantId, string environmentId, string requestId, object body) =>
            ApproveRequestAsync(tenantId, environmentId, requestId, body).GetAwaiter().GetResult();

        public Task<JObject> CancelRequestAsync(string tenantId, string environmentId, string requestId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/elevate/access-requests/{requestId}/cancel");

        public JObject CancelRequest(string tenantId, string environmentId, string requestId) =>
            CancelRequestAsync(tenantId, environmentId, requestId).GetAwaiter().GetResult();

        public Task<JObject> DenyRequestAsync(string tenantId, string environmentId, string requestId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/elevate/access-requests/{requestId}/deny",
                body);

        public JObject DenyRequest(string tenantId, string environmentId, string requestId, object body) =>
            DenyRequestAsync(tenantId, environmentId, requestId, body).GetAwaiter().GetResult();

        public Task<JObject> GetPolicyAsync(string tenantId, string environmentId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"{AuthdogClient.Env(tenantId, environmentId)}/elevate/policy");

        public JObject GetPolicy(string tenantId, string environmentId) =>
            GetPolicyAsync(tenantId, environmentId).GetAwaiter().GetResult();

        public Task<JObject> UpdatePolicyAsync(string tenantId, string environmentId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Put,
                $"{AuthdogClient.Env(tenantId, environmentId)}/elevate/policy",
                body);

        public JObject UpdatePolicy(string tenantId, string environmentId, object body) =>
            UpdatePolicyAsync(tenantId, environmentId, body).GetAwaiter().GetResult();
    }
}

File: java/src/main/java/com/authdog/resources/ElevateResource.java
package com.authdog.resources;

import com.authdog.AuthdogClient;
import com.authdog.exceptions.ApiException;
import com.authdog.exceptions.AuthenticationException;
import com.fasterxml.jackson.databind.JsonNode;
import java.util.Map;

/**
 * Privileged access elevate operations.
 */
public final class ElevateResource {
    /**
     * Authdog client.
     */
    private final AuthdogClient client;

    /**
     * Create an elevate helper.
     * @param clientParam Authdog client
     */
    public ElevateResource(final AuthdogClient clientParam) {
        this.client = clientParam;
    }

    /**
     * Activate an access grant.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param grantId grant ID
     * @param body request body
     * @return activate envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode activateGrant(final String tenantId,
                                  final String environmentId,
                                  final String grantId,
                                  final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/elevate/access-grants/" + grantId
                        + "/activate",
                body, null, JsonNode.class);
    }

    /**
     * Revoke an access grant.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param grantId grant ID
     * @param body request body
     * @return revoke envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode revokeGrant(final String tenantId,
                                final String environmentId,
                                final String grantId,
                                final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/elevate/access-grants/" + grantId
                        + "/revoke",
                body, null, JsonNode.class);
    }

    /**
     * List access requests.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return requests envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode listRequests(final String tenantId,
                                 final String environmentId)
            throws AuthenticationException, ApiException {
        return listRequests(tenantId, environmentId, null);
    }

    /**
     * List access requests filtered by status.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param status optional status filter
     * @return requests envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode listRequests(final String tenantId,
                                 final String environmentId,
                                 final String status)
            throws AuthenticationException, ApiException {
        final Map<String, String> params = QueryParams.create();
        QueryParams.put(params, "status", status);
        return client.request("GET",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/elevate/access-requests",
                null, QueryParams.orNull(params), JsonNode.class);
    }

    /**
     * Create an access request.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param body request body
     * @return created request envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode createRequest(final String tenantId,
                                  final String environmentId,
                                  final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/elevate/access-requests",
                body, null, JsonNode.class);
    }

    /**
     * Get an access request.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param requestId request ID
     * @return request envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode getRequest(final String tenantId,
                               final String environmentId,
                               final String requestId)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/elevate/access-requests/" + requestId,
                null, null, JsonNode.class);
    }

    /**
     * Approve an access request.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param requestId request ID
     * @param body request body
     * @return approve envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode approveRequest(final String tenantId,
                                   final String environmentId,
                                   final String requestId,
                                   final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/elevate/access-requests/" + requestId
                        + "/approve",
                body, null, JsonNode.class);
    }

    /**
     * Cancel an access request.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param requestId request ID
     * @return cancel envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode cancelRequest(final String tenantId,
                                  final String environmentId,
                                  final String requestId)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/elevate/access-requests/" + requestId
                        + "/cancel",
                null, null, JsonNode.class);
    }

    /**
     * Deny an access request.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param requestId request ID
     * @param body request body
     * @return deny envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode denyRequest(final String tenantId,
                                final String environmentId,
                                final String requestId,
                                final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/elevate/access-requests/" + requestId
                        + "/deny",
                body, null, JsonNode.class);
    }

    /**
     * Get the elevate policy.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return policy envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode getPolicy(final String tenantId,
                              final String environmentId)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/elevate/policy",
                null, null, JsonNode.class);
    }

    /**
     * Update the elevate policy.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param body request body
     * @return policy envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode updatePolicy(final String tenantId,
                                 final String environmentId,
                                 final Object body)
            throws AuthenticationException, ApiException {
        return client.request("PUT",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/elevate/policy",
                body, null, JsonNode.class);
    }
}
